packages/fxa-auth-server/fxa-oauth-server/test/api.js

```diff
@@ -4771,6 +4816,380 @@ describe('/v1', function() {
+        validators.jwt.validate(ppidRefreshTokenResult.result.access_token)
+          .error
+      );
+
+      const ppidRefreshTokenJWT = decodeJWT(
+        ppidRefreshTokenResult.result.access_token
+      );
+
+      assert.strictEqual(
+        ppidJWT.claims.sub,
+        ppidRefreshTokenJWT.claims.sub
+      );
+    });
+  });
+
+  describe('Rotating PPIDs', () => {
+    let rotatingSubClient;
+    let rotatingSubClientId;
+
+    beforeEach(() => {
+      rotatingSubClient = clientByName('Rotating PPID JWT Client');
+      rotatingSubClientId = rotatingSubClient.id;
+    });
+
+    it('does not require a ppid_seed', async () => {
+      mockAssertion().reply(200, VERIFY_GOOD);
+
+      const tokenResult = await newToken(
+        {
+          access_type: 'offline',
+        },
+        {
+          clientId: rotatingSubClientId
+        }
+      );
+      assert.equal(tokenResult.statusCode, 200);
+      assert.isNull(
+        validators.jwt.validate(tokenResult.result.access_token).error
+      );
+      const tokenJWT = decodeJWT(tokenResult.result.access_token);
+
+      const sameSeedRefreshTokenResult = await Server.api.post({
+        url: '/token',
+        payload: {
+          client_id: rotatingSubClientId,
+          client_secret: secret,
+          grant_type: 'refresh_token',
+          refresh_token: tokenResult.result.refresh_token,
+        },
+      });
+
+      assert.equal(sameSeedRefreshTokenResult.statusCode, 200);
+      assert.isNull(
+        validators.jwt.validate(sameSeedRefreshTokenResult.result.access_token).error
+      );
+      const sameSeedJWT = decodeJWT(sameSeedRefreshTokenResult.result.access_token);
+
+      assert.strictEqual(sameSeedJWT.claims.sub, tokenJWT.claims.sub);
+    });
+
+    it('accepts ppid_seed when trading code for a token', async () => {
+      mockAssertion().reply(200, VERIFY_GOOD);
+
+      const tokenResult = await newToken(
+        {
+          access_type: 'offline',
+        },
+        {
+          clientId: rotatingSubClientId,
+          ppidSeed: 100
+        }
+      );
+      assert.equal(tokenResult.statusCode, 200);
+      assert.isNull(
+        validators.jwt.validate(tokenResult.result.access_token).error
+      );
+      const tokenJWT = decodeJWT(tokenResult.result.access_token);
+
+
+      const sameSeedRefreshTokenResult = await Server.api.post({
+        url: '/token',
+        payload: {
+          client_id: rotatingSubClientId,
+          client_secret: secret,
+          grant_type: 'refresh_token',
+          ppid_seed: 100,
+          refresh_token: tokenResult.result.refresh_token,
+        },
+      });
+      assert.equal(sameSeedRefreshTokenResult.statusCode, 200);
+      assert.isNull(
+        validators.jwt.validate(sameSeedRefreshTokenResult.result.access_token).error
+      );
+      const sameSeedJWT = decodeJWT(sameSeedRefreshTokenResult.result.access_token);
+
+      assert.strictEqual(sameSeedJWT.claims.sub, tokenJWT.claims.sub);
```

Comment:
I'm not really happy with this test because the sub could rotate due to server time anyways. I'm unsure of a good way to test this w/o depending on really long delays, and even then it's possible for server clock based rotation.

Reply:
You could consider retrying once if the subs don't match, sleeping until `Date.now()` passes a rotation epoch.  Or perhaps explore mocking out `Date.now`, I don't know if we're ever done that in this repo yet...

Reply:
Mocking out Date.now would be alright for unit tests, for integration tests that seems infeasible.